modules/basicfuncs/tf-template.c

```diff
@@ -68,16 +105,73 @@ tf_template_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent,
       return FALSE;
     }
 
-  return tf_template_lookup_invoked_template(state, parent->cfg, invoked_template_name, error);
+  state->invoked_template = tf_template_lookup_invoked_template(state, parent->cfg, invoked_template_name, error);
+  if (tf_template_statically_bound(state))
+    return TRUE;
+
+  /* compile time lookup failed, let's check if it's a dynamically bound invocation */
+  if (strchr(invoked_template_name, '$') == NULL)
+    {
+      /* our argument is not a template, no chance of being better at
+       * runtime, raise as an error */
+
+      g_set_error(error, LOG_TEMPLATE_ERROR, LOG_TEMPLATE_ERROR_COMPILE,
+                  "$(template) Unknown template function or template \"%s\"",
+                  invoked_template_name);
+      return FALSE;
+    }
+
+  state->cfg = parent->cfg;
+  return tf_simple_func_prepare(self, s, parent, argc, argv, error);
+}
+
+static void
+tf_template_eval(LogTemplateFunction *self, gpointer s, LogTemplateInvokeArgs *args)
+{
+  TFTemplateState *state = (TFTemplateState *) s;
+
+  if (tf_template_dynamically_bound(state))
+    tf_simple_func_eval(self, s, args);
+}
+
+static LogTemplate *
+tf_template_get_template_to_be_invoked(TFTemplateState *state, const LogTemplateInvokeArgs *args)
+{
+  LogTemplate *invoked_template;
+
+  if (tf_template_dynamically_bound(state))
+    {
+      const gchar *template_name = args->argv[0]->str;
+
+      invoked_template = cfg_tree_lookup_template(&state->cfg->tree, template_name);
+
+      msg_trace("$(template) dynamic template lookup result",
+                evt_tag_str("template", template_name),
+                evt_tag_int("found", invoked_template != NULL));
+    }
+  else
+    {
+      invoked_template = log_template_ref(state->invoked_template);
```

Comment:
just a note: we could early return in case of static templates... 
(it's just a pattern we are following recently, but I'm OK with this as-is)